docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+
+```logql
+{job="prod/app"} | logfmt client_ip path
+```
+
+will transform the log line
+
+```log
+time=03-10-2020T20:00:01 path=/foo/bar client_ip="127.0.0.1"
+```
+
+into those fields:
+
+```json
+{
+    "client_ip": "127.0.0.1",
+    "path": "/foo/bar"
+}
+```
+
+Again it might be useful to allow renaming fields to avoid conflicts with labels.
+
+For instance `{job="prod/app"} | logfmt ip=client_ip path` will rename the field `client_ip` into `ip`.
+
+`logfmt` will also support implicit field extraction like `json`.
+
+### Regexp
+
+At last, the regexp operator should cover all other use cases but is probably the less efficient and intuitive.
+
+> While we already support `|~` and `!~` filters, we don't want to mix parsers and filters since their purpose are different, filters should be used to narrow the log volume and reduce the number of lines to parse. This is why we will use the notation `| regexp` for the parser operator.
+
+For example the expression `{job="prod/app"} | regexp "(?<client_ip>.+) - - (?<path>.+)"` will transform the log line `127.0.0.1 - - /foo/bar` into :
+
+```json
+{
+    "client_ip": "127.0.0.1",
+    "path": "/foo/bar"
+}
+```
+
+### Questions
+
+- Should we allow to implicit field extraction such as `{job="prod/app"} | logfmt` and `{job="prod/app"} | json` would extract all fields ? I really like the simplicity of it.
+- In the previous document the `as field_name1, field_name2` field renaming notation was also interesting should we use is over `ip=client_ip` for now I've opted for the later which can be omitted in case of no renaming ?
+- Do we prefer `|json` or `| json` notation ?
+
+## LogQL Fields Operators
+
+Now that we have transformed our log streams into log fields we can pipe/chain a new set of operators.
+(Some operators are also compatible with log stream such as the `limit` and `uniq` operators.)
+
+For example the query below:
+
+```logql
+{job="prod/query-frontend"} | logfmt | filter level="info" and latency > 10 | sort duration desc | limit 25 | select query duration
+```
+
+returns the first 25 query and duration slower than 10s ordered by the slowest in the requested time range.
+
+Let's have a look in details how those new operators work.
+
+### Filter
+
+A filter operation noted `| filter` can filter out fields by their value using test predicates such as `path = "/foo/bar"`.
+
+Multiple test predicates can be chained/combined together using `and` and `or` logic operations.
+
+At least one leg of the test predicate must be a literal, this is to ensure we can infer the type to use for comparison. We could later add support for cross field comparison.
+
+__String literal__ will support the same Prometheus match type. (`=`,`=!`,`!~`,`=~`).
+
+Example: `{job="prod/query-frontend"} | logfmt | filter level=~"info|debug" and query_type!="filter"`
+
+__Numeric literal__ will support `=`,`!=` but also `<`,`<=`,`>`,`>=`.
```

Comment:
How is a numeric field detected? Which numeric notations/formats are supported?

Reply:
We want to support float and integer but we should definitively support some scientific notation. This will be part of the implementation details. Thanks for raising this !